### PR TITLE
 feat(ivy): ICU support for Ivy (compiler side)

### PR DESCRIPTION
**Pull request by AndrewKushnir**
The goal of this PR is to add [I18n ICU](https://angular.io/guide/i18n#pluralization) support to IVY compiler. In order to satisfy the backwards (and forward) compatibility requirement, the logic of producing translation messages (`goog.getMsg` at this moment) mimics existing `ng xi18n` machinery output, specifically:

- each ICU takes a separate translation call/unit
- nested ICUs are *not* presented as top-level translations, they remain a part of the parent ICU
- attribute translation also takes a separate call/unit

The I18n AST (which is used to power `ng xi18n` mechanics) was employed to minimize discrepancies between the `ng xi18n` output and IVY compiler output. Nodes within I18n sections now have references to their canonical representation in i18n AST, which helps maintain output consistency.

**Update (11/8)**
As a part of this PR, the following logic was implemented:

- post-processing for `goog.getMsg` output, since the placeholders might not be unique (for ex. `CLOSE_TAG_DIV` can be used multiple times, but we have to have different references/indices). Right now in such cases we put a specially-formatted string that contains these combinations (for ex. `[�/#1�|�/#4:3�|�7:4�]`). The post-processing step assembles the final string at runtime, based on this information.
- the `ConstanlPool` was refactored and all i18n-related code was moved away
- added support for translations without root element (e.g. `<ng-container i18n>`)
- added support for `<ng-template i18n>`
- support for `goog.getMsg` was improved to generate compatible translation strings and respective placeholders in appropriate format
- implemented `i18n` self-closing instruction (in case there are no elements between `i18nStart` and `i18nEnd`)

This PR now completes the main set of features required for i18n compiler.

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/compiler/src/render3/r3_ast.ts [hunk 1]
```diff
@@ -95,6 +99,14 @@ export class Reference implements Node {
   visit<Result>(visitor: Visitor<Result>): Result { return visitor.visitReference(this); }
 }
 
+export class Expansion implements Node {
```
Comment (trotyl) [inline packages/compiler/src/render3/r3_ast.ts]: Wonder whether the name `Expansion` is still right, previous the ICU block was expanded to normal template as a preprocess step, but this is not the case in Ivy, which is totally handled by standalone instructions.

**Review by mhevery — CHANGES_REQUESTED** [review]
I would like @alxhub to have a look at the compiler details as he is more familiar with it. Left comments on the generated code.

File: packages/compiler/src/render3/view/template.ts [hunk 2]
```diff
@@ -1466,6 +1534,13 @@ export function parseTemplate(
   }
 
   let rootNodes: html.Node[] = parseResult.rootNodes;
+
+  // process i18n meta information (scan attributes, generate ids)
+  // before we run whitespace removal process, because existing i18n
+  // extraction process (ng xi18n) relies on a raw content to generate
+  // message ids
+  rootNodes = html.visitAll(new I18nMetaVisitor(), rootNodes);
```
Comment (ocombe) [inline packages/compiler/src/render3/view/template.ts]: While this is true, I believe that the extraction process should also remove whitespace (except if the option `preserveWhitespaces` is set to true). This is a breaking change but since you can use the flag to ensure backwards compatibility, I think that we should make it.

File: packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts [hunk 3]
```diff
@@ -908,6 +996,466 @@ describe('i18n support in the view compiler', () => {
...
+        /**
+         * @desc [BACKUP_MESSAGE_ID:7842238767399919809]
+         */
+        const $MSG_APP_SPEC_TS_1$ = goog.getMsg("{VAR_SELECT, select, male {male} female {female} other {other}}", { VAR_SELECT: "\uFFFD0\uFFFD" });
+        …
+        template: function MyComponent_Template(rf, ctx) {
+          if (rf & 1) {
+            $r3$.ɵelementStart(0, "div");
+            $r3$.ɵi18nStart(1, $MSG_APP_SPEC_TS_1$);
+            $r3$.ɵi18nEnd();
```
Comment (ocombe) [inline packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts]: should be `$r3$.ɵi18n();` instead of "start" and "end"

**Review by ocombe — CHANGES_REQUESTED** [review]
A few things to change, see my inline comments.
And check with @mhevery if we should apply the whitespace removal, and also if we should keep empty translations or not.

**Comment by ocombe** [comment]
Oh and don't forget to update STATUS.md to check "ICU" for the compiler too

File: packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts [hunk 4]
```diff
@@ -908,6 +996,466 @@ describe('i18n support in the view compiler', () => {
...
+            $r3$.ɵelementEnd();
+          }
+          if (rf & 2) {
+            $r3$.ɵi18nExp($r3$.ɵbind(ctx.gender));
+            $r3$.ɵi18nExp($r3$.ɵbind(((ctx.ageA + ctx.ageB) + ctx.ageC)));
```
[the remark this answers is not in the document]
Comment (AndrewKushnir) [reply packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts]: Discussed with Olivier. The binding order can remain as is.

File: packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts [hunk 5]
```diff
@@ -910,6 +964,353 @@ describe('i18n support in the view compiler', () => {
...
+            export class MyModule {}
+          `
+        }
+      };
+
+      const template = String.raw `
+        /**
+         * @desc [BACKUP_MESSAGE_ID:7842238767399919809]
+         */
+        const $MSG_APP_SPEC_TS_1$ = goog.getMsg("{VAR_SELECT, select, male {male} female {female} other {other}}", { VAR_SELECT: "\uFFFD0\uFFFD" });
```
[the remark this answers is not in the document]
Comment (AndrewKushnir) [reply packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts]: I've added post-processing step for ICUs.

File: packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts [hunk 6]
```diff
@@ -910,6 +964,353 @@ describe('i18n support in the view compiler', () => {
...
+         */
+        const $MSG_APP_SPEC_TS_1$ = goog.getMsg("{VAR_SELECT, select, male {male} female {female} other {other}}");
+        /**
+         * @desc [BACKUP_MESSAGE_ID:7068143081688428291]
+         */
+        const $MSG_APP_SPEC_TS_3$ = goog.getMsg("{VAR_SELECT, select, 10 {ten} 20 {twenty} 30 {thirty} other {other}}");
+        /**
+         * @desc [BACKUP_MESSAGE_ID:7254457852348609858]
+         */
+        const $MSG_APP_SPEC_TS_5$ = goog.getMsg("\n  {$ICU}\n  {$ICU_1}\n", { ICU: $r3$.ɵi18nIcuReplaceVars($MSG_APP_SPEC_TS_1$, { VAR_SELECT: "\uFFFD0\uFFFD" }), ICU_1: $r3$.ɵi18nIcuReplaceVars($MSG_APP_SPEC_TS_3$, { VAR_SELECT: "\uFFFD1\uFFFD" }) });
```
[the remark this answers is not in the document]
Comment (AndrewKushnir) [reply packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts]: The post process was implemented.

File: packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts [hunk 7]
```diff
@@ -756,23 +838,26 @@ describe('i18n support in the view compiler', () => {
 
       const template = String.raw `
         const $_c0$ = [1, "ngIf"];
-        const $MSG_APP_SPEC_TS__1$ = goog.getMsg("Some other content \uFFFD0\uFFFD\uFFFD#3\uFFFDMore nested levels with bindings \uFFFD1\uFFFD\uFFFD/#3\uFFFD");
+        /**
+         * @desc [BACKUP_MESSAGE_ID:3458136581072024741]
+         */
+        const $MSG_APP_SPEC_TS__2$ = goog.getMsg("\n      Some other content {$INTERPOLATION}\n      {$START_TAG_DIV}\n        More nested levels with bindings {$INTERPOLATION_1}\n      {$CLOSE_TAG_DIV}\n    ", { INTERPOLATION: "\uFFFD0\uFFFD", START_TAG_DIV: "\uFFFD#3\uFFFD", INTERPOLATION_1: "\uFFFD1\uFFFD", CLOSE_TAG_DIV: "\uFFFD/#3\uFFFD" });
```
[the remark this answers is not in the document]
Comment (AndrewKushnir) [reply packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts]: Correct, I've added the necessary logic to respect the `preserveWhitespaces` config flag.

**Comment by AndrewKushnir** [comment]
@ocombe @mhevery thanks for the initial review!
I've addressed your comments (and added some additional features). I plan to add extra tests and docs for utils functions, but generally this PR is ready for another review, when you get a chance.
Thank you.

File: packages/compiler/src/constant_pool.ts [hunk 8]
```diff
@@ -149,23 +127,27 @@ export class ConstantPool {
...
-    // The identity of an i18n message depends on the message and its meaning
-    const key = parsedMeta.meaning ? `${message}\u0000\u0000${parsedMeta.meaning}` : message;
-
-    const exp = this.translations.get(key);
-
-    if (exp) {
-      return exp;
+  getTranslation(
+      message: string, meta: I18nMeta, params: {[name: string]: o.Expression},
+      ref: {suffix: string, exp?: o.ReadVarExpr}, transformFn?: any): o.Expression {
```
Comment (alxhub) [inline packages/compiler/src/constant_pool.ts]: A couple points:

* Is there a better type for `transformFn` than `any`?
* I think we should not have `exp` optional (for `ref`). I think we should have one way to use the API, and you should always call `freshTranslationName` yourself if you need a `ReadVarExpr`.
* We should rename this function. The original name was bad, but in general functions prefixed with `get` I would _not_ expect mutate internal state unless it's clear from the name that they do (e.g. `getNextAvailableId`).
* I don't think this code belongs in `ConstantPool` - we should move translations out entirely if possible. The only things the `ConstantPool` should be used for are 1) allocation of the variable names and 2) tracking the eventual constant declarations. The _values_ of those constants shouldn't ever be known to the `ConstantPool`.

File: packages/compiler/src/i18n/i18n_ast.ts [hunk 9]
```diff
@@ -69,7 +69,7 @@ export class Icu implements Node {
   public expressionPlaceholder !: string;
   constructor(
       public expression: string, public type: string, public cases: {[k: string]: Node},
-      public sourceSpan: ParseSourceSpan) {}
+      public sourceSpan: ParseSourceSpan, public name?: string) {}
```
Comment (alxhub) [inline packages/compiler/src/i18n/i18n_ast.ts]: What is the `name`? Why is it optional?

File: packages/compiler/src/i18n/i18n_parser.ts [hunk 10]
```diff
@@ -40,25 +44,36 @@ class _I18nVisitor implements html.Visitor {
   private _placeholderToContent !: {[phName: string]: string};
   // TODO(issue/24571): remove '!'.
   private _placeholderToMessage !: {[phName: string]: i18n.Message};
+  // TODO(issue/24571): remove '!'.
+  private _visitNodeFn !: VisitNodeFn | undefined;
```
Comment (alxhub) [inline packages/compiler/src/i18n/i18n_parser.ts]: No need for the TODO or the `!`. Making the field `|undefined` means that it's implicitly initialized to `undefined` (and that's legal).

In general these TODOs are for us to clean things up in the codebase, we should never be adding new instances of them.

File: packages/compiler/src/render3/r3_identifiers.ts [hunk 11]
```diff
@@ -35,9 +35,11 @@ export class Identifiers {
 
   static elementClassProp: o.ExternalReference = {name: 'ɵelementClassProp', moduleName: CORE};
 
-  static elementContainerStart: o.ExternalReference = {name: 'ɵEC', moduleName: CORE};
```
Comment (alxhub) [inline packages/compiler/src/render3/r3_identifiers.ts]: This seems unrelated to i18n and ICUs. It should probably be in a separate commit.

File: packages/compiler/src/render3/r3_template_transform.ts [hunk 12]
```diff
@@ -175,61 +185,78 @@ class HtmlAstToIvyAst implements html.Visitor {
...
-    const valueNoNgsp = replaceNgsp(text.value);
-    const expr = this.bindingParser.parseInterpolation(valueNoNgsp, text.sourceSpan);
-    return expr ? new t.BoundText(expr, text.sourceSpan) : new t.Text(valueNoNgsp, text.sourceSpan);
+    return this._visitTextWithInterpolation(text.value, text.sourceSpan, text.i18n);
   }
 
-  visitComment(comment: html.Comment): null { return null; }
-
-  visitExpansion(expansion: html.Expansion): null { return null; }
+  visitExpansion(expansion: html.Expansion): t.Expansion {
```
Comment (alxhub) [inline packages/compiler/src/render3/r3_template_transform.ts]: Nice!

File: packages/compiler/src/render3/view/i18n/context.ts [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+
+  constructor(
+      private index: number, private ref: o.ReadVarExpr, private level: number = 0,
+      private templateIndex: number|null = null, private meta: I18nMeta,
+      private registryInstance?: any) {
+    this.registry = registryInstance || this.setupRegistry();
+    this.id = this.registry.getUniqueId();
+  }
+
+  private setupRegistry() {
```
Comment (alxhub) [inline packages/compiler/src/render3/view/i18n/context.ts]: Does not need to be a class method.

File: packages/compiler/src/render3/view/i18n/context.ts [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+    const {ast} = element.i18n;
+    if (ast.isVoid && closed) {
+      return;  // ignore "close" for void tags
+    }
+    const ph = ast.isVoid || !closed ? ast.startName : ast.closeName;
+    const content = {type, index, ctx: this.id, isVoid: ast.isVoid, closed};
+    updatePlaceholderMap(this.placeholders, ph, content);
+  }
+
+  getId() { return this.id; }
```
Comment (alxhub) [inline packages/compiler/src/render3/view/i18n/context.ts]: These fields can be public/`readonly` instead of having getters.

File: packages/compiler/src/render3/view/i18n/context.ts [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+  getRef() { return this.ref; }
+  getIndex() { return this.index; }
+  getMeta() { return this.meta; }
+  getIcus() { return this.registry.icus; }
```
Comment (alxhub) [inline packages/compiler/src/render3/view/i18n/context.ts]: Perhaps this should be a getter instead of a `getX()` function.

File: packages/compiler/src/render3/view/i18n/context.ts [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+  getBindings() { return this.bindings; }
+  getRawPlaceholders() { return this.placeholders; }
+  getPlaceholders() {
+    const result = new Map<string, any[]>();
+    this.placeholders.forEach(
+        (values, key) => result.set(key, values.map(serializePlaceholderValue)));
+    return result;
+  }
+
+  isRoot() { return this.level === 0; }
```
Comment (alxhub) [inline packages/compiler/src/render3/view/i18n/context.ts]: Getters :)

File: packages/compiler/src/render3/view/i18n/context.ts [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+  isResolved() { return this.unresolvedCtxCount === 0; }
+
+  appendBinding(binding: o.Expression) { this.bindings.add(binding); }
+  appendIcu(expansion: t.Expansion, ref: o.Expression) {
```
Comment (alxhub) [inline packages/compiler/src/render3/view/i18n/context.ts]: In general, it's considered [better design](https://en.wikipedia.org/wiki/Law_of_Demeter) to take the values you need in parameters directly, instead of accepting a parameter of a given type and only using it to access some field on the parameter.

This would avoid the need to assert `expansion.i18n` is not null here, in order to access `expansion.i18n.ast`.

File: packages/compiler/src/render3/view/i18n/context.ts [hunk 18]
```diff
@@ -0,0 +1,187 @@
...
+    const {name} = expansion.i18n !.ast;
+    updatePlaceholderMap(this.registry.icus, name, ref);
+  }
+  appendBoundText(element: t.BoundText) {
```
Comment (alxhub) [inline packages/compiler/src/render3/view/i18n/context.ts]: The same goes for `element` here.

File: packages/compiler/src/render3/view/i18n/context.ts [hunk 19]
```diff
@@ -0,0 +1,187 @@
...
+    // since we process nested templates separately
+    this.appendTag(TagType.TEMPLATE, element, index, false);
+    this.appendTag(TagType.TEMPLATE, element, index, true);
+    this.unresolvedCtxCount++;
+  }
+  appendElement(element: t.Element, index: number, closed?: boolean) {
+    this.appendTag(TagType.ELEMENT, element, index, closed);
+  }
+
+  forkChildContext(index: number, templateIndex: number, meta: I18nMeta) {
```
Comment (alxhub) [inline packages/compiler/src/render3/view/i18n/context.ts]: This function could use a comment explaining what it does and under what circumstances it might be called.

File: packages/compiler/src/render3/view/i18n/context.ts [hunk 20]
```diff
@@ -0,0 +1,187 @@
...
+    return new I18nContext(index, this.ref, this.level + 1, templateIndex, meta, this.registry);
+  }
+  reconcileChildContext(context: I18nContext) {
```
Comment (alxhub) [inline packages/compiler/src/render3/view/i18n/context.ts]: And here.

File: packages/compiler/src/render3/view/i18n/meta.ts [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+ * i18n-related attributes ("i18n" and "i18n-*") into i18n meta object that is
+ * stored with other element's and atrribute's information.
+ */
+export class I18nMetaVisitor implements html.Visitor {
+  // i18n message generation factory
+  private _createI18nMessage = createI18nMessageFactory(DEFAULT_INTERPOLATION_CONFIG);
+
+  constructor(private config: {keepI18nAttrs: boolean}) {}
+
+  private _setI18nRefs(html: any, i18n: i18n.Node) {
```
Comment (alxhub) [inline packages/compiler/src/render3/view/i18n/meta.ts]: `html` needs a better type than `any` - otherwise how do we know the assignment of a default value of `{}` is safe to `html.i18n`?

This function should not be a private instance method, it doesn't access anything in `I18nMessageVisitor`.

File: packages/compiler/src/render3/view/template.ts [hunk 22]
```diff
@@ -290,11 +376,23 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
...
+      this.creationInstruction(span, R3.i18nEnd);
+    }
     this.i18n = null;  // reset local i18n context
   }
 
+  isSingleElementTemplate(children: t.Node[]): boolean {
+    return children.length === 1 && children[0] instanceof t.Element;
+  }
+
+  hasTextChildrenOnly(children: t.Node[]): boolean {
```
Comment (alxhub) [inline packages/compiler/src/render3/view/template.ts]: This doesn't need to be an instance method.

File: packages/compiler/src/render3/view/template.ts [hunk 23]
```diff
@@ -290,11 +376,23 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
...
     }
-
-    this.creationInstruction(span, R3.i18nEnd);
+    if (!selfClosing) {
+      this.creationInstruction(span, R3.i18nEnd);
+    }
     this.i18n = null;  // reset local i18n context
   }
 
+  isSingleElementTemplate(children: t.Node[]): boolean {
```
Comment (alxhub) [inline packages/compiler/src/render3/view/template.ts]: This doesn't need to be an instance method.

Also, you could write this signature as:

```typescript
function isSingleElementTemplate(children: any[]): children is [t.Element]
```

which would allow type inference to work inside `if` blocks using this function, and eliminate the need for the `elements[0] as ts.Element` cast above.

File: packages/compiler/src/render3/view/template.ts [hunk 24]
```diff
@@ -339,32 +437,26 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
     const elementIndex = this.allocateDataSlot();
 
     let isNonBindableMode: boolean = false;
-    let isI18nRootElement: boolean = false;
+    let isI18nRootElement: boolean = isI18nRootNode(element.i18n);
```
Comment (alxhub) [inline packages/compiler/src/render3/view/template.ts]: This no longer needs to be a `let` but can be `const`.

**Review by alxhub — CHANGES_REQUESTED** [review]
First round of comments.

In general, I notice that `<element>.i18n !` happens a lot - we're constantly asserting in this PR that `i18n` fields are not null. I think this indicates a more fundamental issue with the typings. Can you think of any way to avoid this? I've not really studied the problem deeply but I do have one observation which might be helpful.

Part of the problem comes up at the boundary between i18n-processing code and the larger `TemplateDefinitionBuilder`. We do things like:

```typescript
if (element.i18n) {
 // i18n field is known to be defined here.
 doSomethingWithElement(element);
}

function doSomethingWithElement(element: t.Element) {
  // element.i18n is not known to be defined, so assertion is required.
  console.log(element.i18n !);
}
```

Instead of this pattern, we could make functions on the boundary to which i18n-enabled nodes are passed have signatures like:

```typescript
function doSomethingWithElement(element: t.Element & {i18n: I18nType})
```

This way:
1) we can only call the function on a `t.Element` with a valid `i18n` field.
2) we don't lose that information in the body of the function, and thus don't need the assertion.

This gets the type system working with us and not against us.

Make sense?

File: packages/compiler/src/i18n/i18n_parser.ts [hunk 25]
```diff
@@ -40,25 +44,35 @@ class _I18nVisitor implements html.Visitor {
   private _placeholderToContent !: {[phName: string]: string};
   // TODO(issue/24571): remove '!'.
   private _placeholderToMessage !: {[phName: string]: i18n.Message};
+  private _visitNodeFn !: VisitNodeFn | undefined;
```
Comment (alfaproject) [inline packages/compiler/src/i18n/i18n_parser.ts]: `_visitNodeFn` can be undefined, so I guess you forgot to remove the `!`

File: packages/compiler/src/i18n/i18n_ast.ts (hunk 9 above)
Comment (AndrewKushnir) [reply packages/compiler/src/i18n/i18n_ast.ts]: That was a part of the intermediate change that is not needed anymore, thanks for pointing it out!

File: packages/compiler/src/i18n/i18n_parser.ts (hunk 25 above)
Comment (AndrewKushnir) [reply packages/compiler/src/i18n/i18n_parser.ts]: Yes, thanks for pointing it out!

File: packages/compiler/src/render3/r3_ast.ts (hunk 1 above)
Comment (AndrewKushnir) [reply packages/compiler/src/render3/r3_ast.ts]: Thanks @trotyl, makes sense. I did rename Expansion to Icu in IVY-related code.

File: packages/core/src/render3/i18n.ts [hunk 26]
```diff
@@ -288,14 +294,73 @@ function appendI18nNode(tNode: TNode, parentTNode: TNode, previousTNode: TNode):
   return tNode;
 }
 
-export function i18nAttribute(index: number, attrs: any[]): void {
+export function i18nPostprocess(
```
Comment (ocombe) [inline packages/core/src/render3/i18n.ts]: It needs some API docs because this is exported

File: packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts [hunk 27]
```diff
@@ -905,32 +1038,928 @@ describe('i18n support in the view compiler', () => {
...
+      const input = `
+        <div i18n>{age, select, 10 {ten} 20 {twenty} other {other}}</div>
+      `;
+
+      const output = String.raw `
+        /**
+         * @desc [BACKUP_MESSAGE_ID:8806993169187953163]
+         */
+        const $MSG_APP_SPEC_TS_0_RAW$ = goog.getMsg("{VAR_SELECT, select, 10 {ten} 20 {twenty} other {other}}");
+        const $MSG_APP_SPEC_TS_0$ = $r3$.ɵi18nPostprocess($MSG_APP_SPEC_TS_0_RAW$, { VAR_SELECT: "\uFFFD0\uFFFD" });
```
[the remark this answers is not in the document]
Comment (ocombe) [reply packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts]: I see that you wrote the function and the tests for it below, perfect then.

**Review by ocombe — CHANGES_REQUESTED** [review]

File: packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts [hunk 28]
```diff
@@ -460,109 +539,116 @@ describe('i18n support in the view compiler', () => {
...
+    it('should support named interpolations', () => {
+      const input = `
+        <div i18n>Some value: {{ valueA // i18n(ph="PH_A") }}</div>
+      `;
+
+      const output = String.raw `
+        /**
+         * @desc [BACKUP_MESSAGE_ID:2817319788724342848]
+         */
+        const $MSG_APP_SPEC_TS_0$ = goog.getMsg("Some value: {$phA}", { phA: "\uFFFD0\uFFFD" });
```
[the remark this answers is not in the document]
Comment (AndrewKushnir) [reply packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts]: No, that's the format that `goog.getMsg` enforces, so we convert placeholder names (i..e their format is different in XMB and here, but eventually at merge step they would resolve into the same value).

File: packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts [hunk 29]
```diff
@@ -872,6 +988,23 @@ describe('i18n support in the view compiler', () => {
             $r3$.ɵi18nApply(0);
           }
         }
+        /**
+         * @desc [BACKUP_MESSAGE_ID:1221890473527419724]
+         */
+        const $MSG_APP_SPEC_TS_0_RAW$ = goog.getMsg(" Some content {$startTagDiv_2} Some other content {$interpolation} {$startTagDiv} More nested levels with bindings {$interpolation_1} {$startTagDiv_1} Content inside sub-template {$interpolation_2} {$startTagDiv} Bottom level element {$interpolation_3} {$closeTagDiv}{$closeTagDiv}{$closeTagDiv}{$closeTagDiv}{$startTagDiv_3} Some other content {$interpolation_4} {$startTagDiv} More nested levels with bindings {$interpolation_5} {$closeTagDiv}{$closeTagDiv}", {
+          startTagDiv_2: "\uFFFD*2:1\uFFFD\uFFFD#1:1\uFFFD",
+          closeTagDiv: "[\uFFFD/#2:2\uFFFD|\uFFFD/#1:2\uFFFD\uFFFD/*4:2\uFFFD|\uFFFD/#2:1\uFFFD|\uFFFD/#1:1\uFFFD\uFFFD/*2:1\uFFFD|\uFFFD/#2:3\uFFFD|\uFFFD/#1:3\uFFFD\uFFFD/*3:3\uFFFD]",
```
[the remark this answers is not in the document]
Comment (AndrewKushnir) [reply packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts]: Yes, the `i18nPostprocess` function resolves this by taking values one by one.

File: packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts [hunk 30]
```diff
@@ -905,32 +1038,928 @@ describe('i18n support in the view compiler', () => {
...
+         * @desc [BACKUP_MESSAGE_ID:7842238767399919809]
+         */
+        const $MSG_APP_SPEC_TS_2_RAW$ = goog.getMsg("{VAR_SELECT, select, male {male} female {female} other {other}}");
+        const $MSG_APP_SPEC_TS_2$ = $r3$.ɵi18nPostprocess($MSG_APP_SPEC_TS_2_RAW$, { VAR_SELECT: "\uFFFD1\uFFFD" });
+        const $_c3$ = [1, "ngIf"];
+        /**
+         * @desc [BACKUP_MESSAGE_ID:7842238767399919809]
+         */
+        const $MSG_APP_SPEC_TS__4_RAW$ = goog.getMsg("{VAR_SELECT, select, male {male} female {female} other {other}}");
+        const $MSG_APP_SPEC_TS__4$ = $r3$.ɵi18nPostprocess($MSG_APP_SPEC_TS__4_RAW$, { VAR_SELECT: "\uFFFD0:1\uFFFD" });
```
[the remark this answers is not in the document]
Comment (AndrewKushnir) [reply packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts]: Yes, I've created a separate ticket to handle that.
It might be a bit tricky, since we need to share translation (i.e. `goog.getMsg` call), but not the result of post-processing, and also take placeholder values (2nd argument) into account.

File: packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts [hunk 31]
```diff
@@ -905,32 +1038,928 @@ describe('i18n support in the view compiler', () => {
...
+            $r3$.ɵi18nExp($r3$.ɵbind(ctx.gender));
+            $r3$.ɵi18nApply(1);
+          }
+        }
+      `;
+
+      verify(input, output);
+    });
+
+    it('should ignore icus generated outside of i18n block', () => {
```
[the remark this answers is not in the document]
Comment (AndrewKushnir) [reply packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts]: As we discussed, we will initiate i18n section for such ICUs, that should ensure better backwards-compatibility. I've created a separate ticket to handle that.

File: packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts [hunk 32]
```diff
@@ -460,109 +539,116 @@ describe('i18n support in the view compiler', () => {
...
-              \`
-            })
-            export class MyComponent {}
-
-            @NgModule({declarations: [MyComponent]})
-            export class MyModule {}
-          `
+    it('should support named interpolations', () => {
+      const input = `
+        <div i18n>Some value: {{ valueA // i18n(ph="PH_A") }}</div>
```
[the remark this answers is not in the document]
Comment (AndrewKushnir) [reply packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts]: Created a ticket to add docs to angular.io.

File: packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts [hunk 33]
```diff
@@ -645,63 +733,70 @@ describe('i18n support in the view compiler', () => {
...
+         */
+        const $MSG_APP_SPEC_TS_1$ = goog.getMsg("Span title {$interpolation} and {$interpolation_1}", {
+          interpolation: "\uFFFD0\uFFFD",
+          interpolation_1: "\uFFFD1\uFFFD"
+        });
+        const $_c2$ = ["title", $MSG_APP_SPEC_TS_1$];
+        /**
+         * @desc [BACKUP_MESSAGE_ID:4446430594603971069]
+         */
+        const $MSG_APP_SPEC_TS_0$ = goog.getMsg(" My i18n block #1 with value: {$interpolation} {$startTagSpan} Plain text in nested element (block #1) {$closeTagSpan}", {
```
[the remark this answers is not in the document]
Comment (AndrewKushnir) [reply packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts]: That's the same text string that is generated for `text(...)` instructions, for ex.:
```
  if (rf & 2) {
    textBinding(1, interpolation1(" My i18n block #1 with value: ", ctx.valueA, " "));
```
So I'd keep it as is for consistency. If we want to remove this whitespace, it should probably be handled on a different level instead.

File: packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts [hunk 34]
```diff
@@ -424,34 +489,48 @@ describe('i18n support in the view compiler', () => {
...
-            export class MyComponent {}
-
-            @NgModule({declarations: [MyComponent]})
-            export class MyModule {}
-          `
-        }
-      };
+      const input = `
+        <div i18n></div>
+        <div i18n>  </div>
```
[the remark this answers is not in the document]
Comment (AndrewKushnir) [reply packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts]: Yes, line breaks are ignored as well. I've extended this test to verify that.

File: packages/core/src/render3/i18n.ts (hunk 26 above)
Comment (AndrewKushnir) [reply packages/core/src/render3/i18n.ts]: Thanks, I've added docs.

**Review by ocombe — APPROVED** [review]
Everything seems good for me, just fix the failing circleCI tasks (needs `gulp format`, and cleanup of the imports apparently)

**Review by mhevery — APPROVED** [review]
With minor fixes ab597103c3

File: packages/compiler/src/render3/r3_identifiers.ts (hunk 11 above)
Comment (alxhub) [reply packages/compiler/src/render3/r3_identifiers.ts]: Ah, I think I was a bit confusing. Please do still update this now, in this PR. Just move the update into a separate git _commit_, as it has nothing to do with what's going on in your i18n commit.

**Review by alxhub — APPROVED** [review]
Approved with one comment still :)

**Comment by mhevery** [comment]
http://test/OCL:221805387:BASE:221805417:1542389339980:6cd3f359

**Comment by AndrewKushnir** [comment]
@alxhub thanks for review!
I rebased this PR from the latest master that now includes the change from a different PR (60800da6c17bb20000a06d5c3374eeedf98235ee), which takes care of `elementContainer`-related changes (so my PR no longer contains this code). Thank you.

**Comment by AndrewKushnir** [comment]
New g3 presubmit: http://test/OCL:221834624:BASE:221834969:1542400404739:98891ab3
